src/Frontend/Core.hs

```diff
@@ -221,52 +235,113 @@ percentVotes idea numVoters vv = {- assert c -} v
           else (numVotes idea vv * 100) `div` numVoters
 
 
--- * render context
-
--- | Contains all the information which is needed to render a user role dependent functionality.
-data RenderContext = RenderContext
-      { _renderContextUser     :: User
-      }
-  deriving (Eq, Read, Show)
-
--- | Calculates the render context for role sensitive page rendering
-renderContext :: (ActionPersist m, ActionUserHandler m) => m RenderContext
-renderContext = RenderContext <$> currentUser
-
-
 -- * pages
 
+data AccessResult
+  = AccessGranted
+  | AccessDenied { _accessDeniedMsg :: ST, _accessDeniedRedirect :: Maybe URL }
+  | AccessDeferred
+
+instance Monoid AccessResult where
+    mempty = AccessGranted
+    AccessGranted `mappend` x = x
+    x `mappend` AccessGranted = x
+    AccessDeferred `mappend` x = x
+    x `mappend` AccessDeferred = x
+    AccessDenied s0 u0 `mappend` AccessDenied s1 u1 =
+        AccessDenied (s0 <> "\n\n" <> s1) (getFirst (First u0 <> First u1))
+
+data AccessInput a
+  = NotLoggedIn
+  | LoggedIn { _authUser :: User, _authPage :: Maybe a }
+
+instance Functor AccessInput where
+    fmap f = \case
+        NotLoggedIn -> NotLoggedIn
+        LoggedIn u mp -> LoggedIn u (f <$> mp)
+
 -- | Defines some properties for pages
 class Page p where
-    isPrivatePage :: proxy p -> Bool
-    isPrivatePage _ = True
+    isAuthorized :: Applicative m => AccessInput p -> m AccessResult
 
     extraPageHeaders  :: p -> Html ()
     extraPageHeaders _ = nil
 
     extraBodyClasses  :: p -> [ST]
     extraBodyClasses _ = nil
 
-instance Page () where
-    isPrivatePage _ = False
+accessGranted :: Applicative m => m AccessResult
+accessGranted = pure AccessGranted
+
+accessDenied :: Applicative m => ST -> Maybe (P.Main 'P.AllowGetPost) -> m AccessResult
+accessDenied m u = pure $ AccessDenied m (absoluteUriPath . relPath <$> u)
+
+accessDeferred :: Applicative m => m AccessResult
+accessDeferred = pure AccessDeferred
+
+redirectLogin :: Applicative m => m AccessResult
+redirectLogin = accessDenied "Not logged in" $ Just P.Login
 
-instance Page ST where
-    isPrivatePage _ = True -- safer default, might need to be changed if needed
+userPage :: Applicative m => AccessInput any -> m AccessResult
+userPage LoggedIn{}  = accessGranted
+userPage NotLoggedIn = redirectLogin
+
+pageForRole :: Applicative m => Role -> AccessInput any -> m AccessResult
+pageForRole r (LoggedIn u _)
+    | u ^. userRole == r  = accessGranted
+    | otherwise           = accessDenied ("Expecting " <> r ^. uilabeled <> " role.") Nothing
+pageForRole _ NotLoggedIn = redirectLogin
+
+publicPage :: Applicative m => any -> m AccessResult
+publicPage _ = accessGranted
+
+adminPage :: Applicative m => AccessInput any -> m AccessResult
+adminPage = pageForRole Admin
+
+authNeedPage :: Applicative m => (p -> m AccessResult) -> AccessInput p -> m AccessResult
+authNeedPage withPage = \case
```

Comment:
accessToPage?

`page-p` is lisp and prolog, but `on` in haskell i think has a different intuition.

since it's just a local name, please make a call and i'll be happy with it.  `withPage` is good enough, if none of the others seem better to you.